rewrite-yaml/src/main/java/org/openrewrite/yaml/MergeYamlVisitor.java

```diff
@@ -89,96 +101,160 @@ public Yaml visitSequence(Yaml.Sequence existingSeq, P p) {
     }
 
-    private boolean keyMatches(Yaml.Mapping m1, Yaml.Mapping m2) {
+    private boolean keyMatches(Mapping m1, Mapping m2) {
         Optional<String> nameToAdd = m2.getEntries().stream()
                 .filter(e -> objectIdentifyingProperty != null && objectIdentifyingProperty.equals(e.getKey().getValue()))
-                .map(e -> ((Yaml.Scalar) e.getValue()).getValue())
+                .map(e -> ((Scalar) e.getValue()).getValue())
                 .findAny();
 
         return nameToAdd.map(nameToAddValue -> m1.getEntries().stream()
                         .filter(e -> objectIdentifyingProperty.equals(e.getKey().getValue()))
-                        .map(e -> ((Yaml.Scalar) e.getValue()).getValue())
+                        .map(e -> ((Scalar) e.getValue()).getValue())
                         .anyMatch(existingName -> existingName.equals(nameToAddValue)))
                 .orElse(false);
     }
 
-    private Yaml.Mapping mergeMapping(Yaml.Mapping m1, Yaml.Mapping m2, P p, Cursor cursor) {
-        List<Yaml.Mapping.Entry> mutatedEntries = ListUtils.map(m1.getEntries(), existingEntry -> {
-            for (Yaml.Mapping.Entry incomingEntry : m2.getEntries()) {
+    private Mapping mergeMapping(Mapping m1, Mapping m2, P p, Cursor cursor) {
+        List<Entry> mergedEntries = map(m1.getEntries(), existingEntry -> {
+            for (Entry incomingEntry : m2.getEntries()) {
                 if (keyMatches(existingEntry, incomingEntry)) {
-                    return existingEntry.withValue((Yaml.Block) new MergeYamlVisitor<>(existingEntry.getValue(),
-                            incomingEntry.getValue(), acceptTheirs, objectIdentifyingProperty, shouldAutoFormat)
-                            .visitNonNull(existingEntry.getValue(), p, new Cursor(cursor, existingEntry)));
+                    return existingEntry.withValue((Yaml.Block)
+                            new MergeYamlVisitor<>(existingEntry.getValue(), incomingEntry.getValue(), acceptTheirs, objectIdentifyingProperty, shouldAutoFormat)
+                                    .visitNonNull(existingEntry.getValue(), p, new Cursor(cursor, existingEntry)));
                 }
             }
             return existingEntry;
         });
 
-        mutatedEntries = ListUtils.concatAll(mutatedEntries, ListUtils.map(m2.getEntries(), incomingEntry -> {
-            for (Yaml.Mapping.Entry existingEntry : m1.getEntries()) {
-                if (keyMatches(existingEntry, incomingEntry)) {
+        List<Entry> mutatedEntries = concatAll(mergedEntries, map(m2.getEntries(), it -> {
+            for (Entry existingEntry : m1.getEntries()) {
+                if (keyMatches(existingEntry, it)) {
                     return null;
                 }
             }
-            if (shouldAutoFormat) {
-                incomingEntry = autoFormat(incomingEntry, p, cursor);
+            // workaround: autoFormat put sometimes extra spaces before elements
```

Comment:
Workaround fixes only so many cases; should be removed in the future and done in the parser:
- https://github.com/openrewrite/rewrite/pull/4358